LTS (Gateway & Dashboard) 5.8.9 RNs.mdx (#1019)
TykTechnologies/tyk-docs: developer-support/release-notes/gateway.mdx

|         | Helm Chart v5.0.0 | Helm all versions |
|         | EDP v1.15.0     | EDP all versions |
|         | Pump v1.13.1    | Pump all versions |
|         | TIB (if using standalone) v1.7.0 | TIB all versions |

##### 3rd Party Dependencies & Tools

| Third Party Dependency                                       | Tested Versions        | Compatible Versions    | Comments | 
| :------------------------------------------------------------ | :---------------------- | :---------------------- | :-------- | 
| [Go](https://go.dev/dl/)                                     | 1.24  |  1.24  | [Go plugins](/api-management/plugins/golang) must be built using Go 1.24 | 
| [Redis](https://redis.io/download/)  | 6.2.x, 7.x  | 6.2.x, 7.x  | Used by Tyk Gateway | 
| [OpenAPI Specification](https://spec.openapis.org/oas/v3.0.3)| v3.0.x                 | v3.0.x                 | Supported by [Tyk OAS](/api-management/gateway-config-tyk-oas) |

Given the potential time difference between your upgrade and the release of this version, we recommend users verify the ongoing support of third-party dependencies they install, as their status may have changed since the release.

#### Deprecations

> There are no deprecations in this release.

Reviewer: `allow_explicit_policy_id` is deprecated from this release (per the change log)

We have implemented background monitoring of MDCB endpoint DNS resolution to ensure rapid response to changes without waiting for failures, which block API consumer requests. When a DNS change is detected, Tyk will now automatically reconnect the RPC client to minimise downtime and risk of request blocking. The DNS monitor checks for changes at a configurable interval (default: 30 seconds, minimum: 10 seconds). This can be set using the `slave_options.dns_monitor` configuration.
</Accordion>

</AccordionGroup>

##### Fixed

<AccordionGroup>

<Accordion title='Fixed JWT Authentication Panic in MDCB Emergency Mode'>
Fixed a bug causing JWT authentication to panic in MDCB emergency mode. When processing tokens with new sub-claims, the gateway returned an uninitialized session missing its KeyID, leading to a crash when generating the session’s KeyHash. The fix ensures the KeyID is preserved in the emergency-mode path, allowing sessions to be created and cached correctly and preventing panics during MDCB outages.
</Accordion>

<Accordion title='Resolved Panic Triggered by DRL Updates in Mixed Rate-Limiter Environments'>

> Fixed an issue where Gateways using Redis-based rate limiters would crash when sharing Redis with Gateways using Distributed Rate Limiting (DRL). Non-DRL Gateways now properly ignore DRL update messages instead of attempting to process them, enabling mixed rate-limiter deployments across shared Redis instances.

Reviewer: ```suggestion
Fixed an issue where Gateways using [Redis Rate Limiters](/api-management/rate-limit#redis-rate-limiter) would crash when sharing Redis with Gateways using [Distributed Rate Limiting](/api-management/rate-limit#distributed-rate-limiter) (DRL). Non-DRL Gateways now properly ignore DRL update messages instead of attempting to process them, enabling mixed rate-limiter deployments across shared Redis instances.
```

</Accordion>

<Accordion title='Removed Redundant Boolean Enums from OpenAPI Specification'>
Fixed redundant boolean enum definitions in OpenAPI specification by removing unnecessary enum: [true, false] declarations from boolean type fields in swagger.yml files. Boolean parameters now use only type: boolean, following OpenAPI best practices.
</Accordion>

> <Accordion title='Fixed Data Plane Startup Failures Causing Incorrect 404 Responses'>

Reviewer: ```suggestion
<Accordion title='Fixed Data Plane Startup Process To Handle MDCB Failure>
```

> Fixed an issue where a Data Plane Gateway could fail to load API definitions the MDCB link failed during initialisation. This would lead to client requests returning HTTP 404 errors. The expected behaviour, if MDCB is unavailable, is for the Data Plane Gateway to retrieve policies and API definitions from the local storage (Redis), but this was not occurring in certain scenarios. We have improved the robustness of the Gateway startup so that, if MDCB goes down, it will automatically switch to the local storage (Redis) as expected.

Reviewer: ```suggestion
Fixed an issue where a Data Plane Gateway could return HTTP 404 errors to client requests if the MDCB link failed during initialisation. The expected behaviour, if MDCB is unavailable, is for the Data Plane Gateway to retrieve policies and API definitions from the local storage (Redis), but this was not occurring in certain scenarios. We have improved the robustness of the Gateway startup so that, if MDCB goes down, it will automatically switch to the local storage (Redis) as expected.
```

</Accordion>

<Accordion title='Fixed OAuth Client Key Retrieval Delays in JWT APIs on Hybrid Gateways'> 

> Resolved an issue where JWT APIs using Keycloak authentication experienced significant delays on hybrid gateways due to failed local key lookups. The gateway was unable to find OAuth client keys in local Redis and had to fetch them from the control plane on every request, causing performance degradation and "key not found" errors in logs. JWT API requests now retrieve keys efficiently from local storage, eliminating unnecessary round-trip requests and providing consistent response times. </Accordion>

Reviewer: ```suggestion
Resolved an issue where JWT APIs using Keycloak authentication experienced significant delays on hybrid gateways due to failed local key lookups. The Gateway was unable to find OAuth client keys in local Redis and had to fetch them from the Control Plane for every request, causing performance degradation and "key not found" errors in logs. JWT API requests now retrieve keys efficiently from local storage, eliminating unnecessary round-trip requests and providing consistent response times. </Accordion>
```

Reviewer 2: Added

> 

Reviewer 2: ```suggestion
</Accordion>

```

<Accordion title='Corrected mTLS Certificate Advertising for RFC-Compliant Clients'>
Fixed an issue where Tyk Gateway advertised leaf certificate Subject DNs instead of Certificate Authority DNs during mTLS handshakes, causing connection failures with RFC-compliant TLS clients. The Gateway now properly extracts and advertises CA DNs from certificate chains in the CertificateRequest message, ensuring compatibility with standards-compliant clients like `Apache mod_ssl` while maintaining backward compatibility with existing configurations.
</Accordion>

<Accordion title='Fixed JSON Formatter Failures with Large Numeric Error Values'>
We fixed a logging bug in the JSON formatter that could cause error logs to fail to serialize when an error message contained very large numeric values (for example a big integer), which sometimes resulted in missing or broken log output; the formatter now writes the error as a plain text string instead of attempting to encode the underlying error object, so logs reliably serialize to JSON
</Accordion>

> <Accordion title='Request pipeline blocked by synchronous RPC calls every 10 minutes when MDCB is unavailable'>

Reviewer: ```suggestion
<Accordion title='Fixed Request Pipeline Blockage When MDCB is Unavailable'>
```

> Fixed blocking synchronous RPC calls in the request pipeline that occurred every 10 minutes during organization expiry checks when MDCB was unavailable. The organization expiry validation is now asynchronous and non-blocking, preventing API request timeouts and latency spikes (up to 90 seconds) when MDCB connectivity issues occur. This ensures consistent API response times regardless of MDCB availability status.

Reviewer: ```suggestion
Fixed an issue where the request pipeline could be blocked by synchronous RPC calls when MDCB was unavailable. The Gateway check for Organisation expiry every 10 minutes is now asynchronous and non-blocking, so it no longer causes API request timeouts and latency spikes (up to 90 seconds) when MDCB is unavailable. Previously this validity check would block the Gateway. This fix ensures consistent API response times regardless of MDCB availability.
```

## 5.8 Release Notes

### 5.8.9 Release Notes

> #### Release Date 10 December 2025

Reviewer 2: ```suggestion
#### Release Date 11 December 2025
```

|         | Helm Chart v5.0.0 | Helm all versions |
|         | EDP v1.15.0     | EDP all versions |
|         | Pump v1.13.1    | Pump all versions |
|         | TIB (if using standalone) v1.7.0 | TIB all versions |

##### 3rd Party Dependencies & Tools

| Third Party Dependency                                       | Tested Versions        | Compatible Versions    | Comments | 
| :------------------------------------------------------------ | :---------------------- | :---------------------- | :-------- | 
| [Go](https://go.dev/dl/)                                     | 1.24  |  1.24  | [Go plugins](/api-management/plugins/golang) must be built using Go 1.24 | 
| [Redis](https://redis.io/download/)  | 6.2.x, 7.x  | 6.2.x, 7.x  | Used by Tyk Gateway | 
| [OpenAPI Specification](https://spec.openapis.org/oas/v3.0.3)| v3.0.x                 | v3.0.x                 | Supported by [Tyk OAS](/api-management/gateway-config-tyk-oas) |

Given the potential time difference between your upgrade and the release of this version, we recommend users verify the ongoing support of third-party dependencies they install, as their status may have changed since the release.

#### Deprecations

> In this release we've deprecated `allow_explicit_policy_id`.

Reviewer: ```suggestion
In this release we've deprecated the [policies.allow_explicit_policy_id](/tyk-oss-gateway/configuration#policies-allow-explicit-policy-id) configuration option. This was previously added to allow the use of custom policy IDs, which is now the default behaviour so this option is redundant.
```

</Accordion>

<Accordion title='Removed Redundant Boolean Enums from OpenAPI Specification'>
Fixed redundant boolean enum definitions in OpenAPI specification by removing unnecessary enum: [true, false] declarations from boolean type fields in swagger.yml files. Boolean parameters now use only type: boolean, following OpenAPI best practices.
</Accordion>

<Accordion title='Fixed Data Plane Startup Process To Handle MDCB Failure'>
Fixed an issue where a Data Plane Gateway could return HTTP 404 errors to client requests if the MDCB link failed during initialisation. The expected behaviour, if MDCB is unavailable, is for the Data Plane Gateway to retrieve policies and API definitions from the local storage (Redis), but this was not occurring in certain scenarios. We have improved the robustness of the Gateway startup so that, if MDCB goes down, it will automatically switch to the local storage (Redis) as expected.
</Accordion>

<Accordion title='Fixed OAuth Client Key Retrieval Delays in JWT APIs on Hybrid Gateways'> 

> Resolved an issue where JWT APIs using Keycloak authentication experienced significant delays on hybrid gateways due to failed local key lookups. The gateway was unable to find OAuth client keys in local Redis and had to fetch them from the control plane on every request, causing performance degradation and "key not found" errors in logs. JWT API requests now retrieve keys efficiently from local storage, eliminating unnecessary round-trip requests and providing consistent response times. 

Reviewer 2: ```suggestion
Resolved an issue where JWT APIs using Keycloak authentication experienced significant delays on hybrid gateways due to failed local key lookups. The Gateway was unable to find OAuth client keys in local Redis and had to fetch them from the Control Plane for every request, causing performance degradation and "key not found" errors in logs. JWT API requests now retrieve keys efficiently from local storage, eliminating unnecessary round-trip requests and providing consistent response times. 
```
